src/Dofs/sparsity_pattern.jl

```diff
@@ -1,83 +1,396 @@
+    end
+    return SparsityPattern(nrows, ncols, heap, rows)
+end
+
+function Base.show(io::IO, ::MIME"text/plain", sp::SparsityPattern)
+    iob = IOBuffer()
+    println(iob, "$(n_rows(sp))×$(n_cols(sp)) $(sprint(show, typeof(sp))):")
+    # Collect min/max/avg entries per row
+    min_entries = typemax(Int)
+    max_entries = typemin(Int)
+    stored_entries = 0
+    for r in eachrow(sp)
+        l = length(r)
+        stored_entries += l
+        min_entries = min(min_entries, l)
+        max_entries = max(max_entries, l)
+    end
+    # Print sparsity
+    sparsity_pct = round(
+        (n_rows(sp) * n_cols(sp) - stored_entries) / (n_rows(sp) * n_cols(sp)) * 100 * 1000
+    ) / 1000
+    println(iob, " - Sparsity: $(sparsity_pct)% ($(stored_entries) stored entries)")
+    # Print row stats
+    avg_entries = round(stored_entries / n_rows(sp) * 10) / 10
+    println(iob, " - Entries per row (min, max, avg): $(min_entries), $(max_entries), $(avg_entries)")
+    # Compute memory estimate
+    @assert n_rows(sp) * sizeof(eltype(sp.rows)) == sizeof(sp.rows)
+    bytes_used      = sizeof(sp.rows) + stored_entries * sizeof(Int)
+    bytes_allocated = sizeof(sp.rows) + HeapAllocator.heap_stats(sp.heap)[2]
+    print(iob,   " - Memory estimate: $(Base.format_bytes(bytes_used)) used, $(Base.format_bytes(bytes_allocated)) allocated")
+    write(io, seekstart(iob))
+    return
+end
 
-Create the symmetric sparsity pattern corresponding to the degree of freedom
-numbering in the [`DofHandler`](@ref) by only considering the upper
-triangle of the matrix. Return a `Symmetric{SparseMatrixCSC}`.
+n_rows(sp::SparsityPattern) = sp.nrows
+n_cols(sp::SparsityPattern) = sp.ncols
+
+@inline function add_entry!(sp::SparsityPattern, row::Int, col::Int)
+    @boundscheck 1 <= row <= n_rows(sp) && 1 <= col <= n_cols(sp)
+    r = @inbounds sp.rows[row]
+    r = insert_sorted(r, col)
+    @inbounds sp.rows[row] = r
+    return
+end
+
+@inline function insert_sorted(x::HeapAllocator.HeapVector{Int}, item::Int)
+    k = searchsortedfirst(x, item)
+    if k == length(x) + 1 || @inbounds(x[k]) != item
+        x = HeapAllocator.insert(x, k, item)
+    end
+    return x
+end
 
-See the [Sparsity Pattern](@ref) section of the manual.
+eachrow(sp::SparsityPattern)           = sp.rows
+eachrow(sp::SparsityPattern, row::Int) = sp.rows[row]
+
+
+########################
+# BlockSparsityPattern #
+########################
+
+# This is implemented as an extension in ext/FerriteBlockArrays.jl. This method exist to
+# give a nice error message. Keep the signature in sync with the constructor in
+# ext/FerriteBlockArrays.jl (just a tad less specific here).
+function BlockSparsityPattern(::AbstractVector{<:Integer})
+    msg = "Ferrite's block matrix functionality depends on the package BlockArrays to " *
+          "be installed and loaded into the session. Install BlockArrays " *
+          "(`pkg> add BlockArrays`) and load it (`using BlockArrays`) to enable."
+    error(msg)
+end
```

Comment:
OT: Should I adopt this design for the GPU PR?

Reply:
It is a bit of a hack to export a struct from an extension, not sure if it is a good idea and might move the blocksparsity pattern inside of Ferrite (but keep the matrix instantiation in the extension).

Reply:
I see. Technically we should also be able to just instantiate a normal matrix with a block pattern. Or should this functionality be rather moved into `close!(::DofHandler, ::DofStrategy)``?